### PR TITLE
Add `dtype` to ScaleIntensityRange transforms

### LINKED ISSUE #3234: Add `dtype` arg to `ScaleIntensityRange` transform for output
**Is your feature request related to a problem? Please describe.**
Need to add a `dtype` arg to `ScaleIntensityRange` to control the output datatype, same as other intensity transforms.

### PR DESCRIPTION
**Pull request by Nic-Ma**
Fixes #3234 .

### Description
This PR added `dtype` arg to `ScaleIntensityRange` transforms, same as other `intensity scale` transforms which may change data type during computation, especially for `int` or `long` input.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/transforms/intensity/array.py [hunk 1]
```diff
@@ -748,7 +752,9 @@ def __call__(self, img: NdarrayOrTensor) -> NdarrayOrTensor:
         img = img * (self.b_max - self.b_min) + self.b_min
         if self.clip:
             img = clip(img, self.b_min, self.b_max)
-        return img
+        ret, *_ = convert_data_type(img, dtype=self.dtype)
```
Comment (yiheng-wang-nv) [inline monai/transforms/intensity/array.py:755]: Thanks @Nic-Ma , here should `dtype` be an optional augment?
Then:
```
if self.dtype:
    ret, *_ = convert_data_type(img, dtype=self.dtype)
else:
    ret, *_ = convert_data_type(img, dtype=img.dtype)
```
Some transforms such as `ThresholdIntensity ` does not have this augment, thus if users forget to set the type here, a tensor input may be converted into numpy array.

Comment (Nic-Ma) [reply monai/transforms/intensity/array.py]: Sure, updated all the dtype of transforms in the latest commit.

Thanks.

**Comment by Nic-Ma** [comment]
I also fixed the data in / out type issue in histogram_normalize transform.

Thanks.

**Comment by myron** [comment]
Thanks for this!

**Review by yiheng-wang-nv — APPROVED** [review]

File: monai/transforms/utils.py [hunk 2]
```diff
@@ -149,22 +149,26 @@ def zero_margins(img: np.ndarray, margin: int) -> bool:
...
-
+    dtype_ = dtype or arr.dtype
     mina = arr.min()
     maxa = arr.max()
 
     if mina == maxa:
         return arr * minv
 
     norm = (arr - mina) / (maxa - mina)  # normalize the array first
-    return (norm * (maxv - minv)) + minv  # rescale by minv and maxv, which is the normalized array by default
```
Comment (Nic-Ma) [inline monai/transforms/utils.py:167]: Fixed a bug here that we should convert data type after computation.

File: monai/transforms/intensity/array.py [hunk 3]
```diff
@@ -1988,12 +2007,15 @@ def __init__(
...
+    def __call__(self, img: NdarrayOrTensor, mask: Optional[NdarrayOrTensor] = None) -> NdarrayOrTensor:
+        img_np: np.ndarray
+        img_np, *_ = convert_data_type(img, np.ndarray)  # type: ignore
+        mask = mask if mask is not None else self.mask
+        mask_np: Optional[np.ndarray] = None
+        if mask is not None:
+            mask_np, *_ = convert_data_type(mask, np.ndarray)  # type: ignore
+
+        ret = equalize_hist(img=img_np, mask=mask_np, num_bins=self.num_bins, min=self.min, max=self.max)
+        out, *_ = convert_to_dst_type(src=ret, dst=img, dtype=self.dtype or img.dtype)
```
Comment (Nic-Ma) [inline monai/transforms/intensity/array.py:2019]: Fixed a bug here to support Tensor in -> Tensor out.

Thanks.

**Review by wyli — CHANGES_REQUESTED** [review]
I feel in many cases now it's good to use the `CastToType` to indicate type casting, perhaps this needs more discussion?

**Comment by Nic-Ma** [comment]
> I feel in many cases now it's good to use the `CastToType` to indicate type casting, perhaps this needs more discussion?

Hi @wyli ,

I agree with you, we should use `CastToType` for most cases. Maybe this PR changed many things that confused you.
Actually, here I didn't add `dtype` to more transforms, just added to `ScaleIntensityRange`, because data type may be changed during `scaling`, just followed the same pattern as `ScaleIntensity` transform.
All the other changes are to add support for `dtype=None`.
Could you please help take a look again?

Thanks.

**Review by wyli — APPROVED** [review]
Thanks, we agreed to have this based on the discussions.

File: monai/transforms/utils.py (hunk 2 above)
Comment (wyli) [reply monai/transforms/utils.py:167]: this causes an issue in https://github.com/Project-MONAI/MONAI/pull/3242
```
======================================================================

ERROR: test_blend_2 (tests.test_blend_images.TestBlendImages)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/opt/conda/lib/python3.6/site-packages/parameterized/parameterized.py", line 533, in standalone_func

    return func(*(a + p.args), **p.kwargs)

  File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/test_blend_images.py", line 39, in test_blend

    blended = blend_images(image[None], label[None])

  File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/visualize/utils.py", line 144, in blend_images

    label = rescale_array(label)

  File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/transforms/utils.py", line 167, in rescale_array

    norm = (arr - mina) / (maxa - mina)  # normalize the array first

RuntimeError: Integer division of tensors using div or / is no longer supported, and in a future release div will perform true division as in Python 3. Use true_divide or floor_divide (// in Python) instead.



======================================================================

ERROR: test_blend_3 (tests.test_blend_images.TestBlendImages)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/opt/conda/lib/python3.6/site-packages/parameterized/parameterized.py", line 533, in standalone_func

    return func(*(a + p.args), **p.kwargs)

  File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/test_blend_images.py", line 39, in test_blend

    blended = blend_images(image[None], label[None])

  File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/visualize/utils.py", line 144, in blend_images

    label = rescale_array(label)

  File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/transforms/utils.py", line 167, in rescale_array

    norm = (arr - mina) / (maxa - mina)  # normalize the array first

RuntimeError: Integer division of tensors using div or / is no longer supported, and in a future release div will perform true division as in Python 3. Use true_divide or floor_divide (// in Python) instead.



======================================================================

ERROR: test_blend_4 (tests.test_blend_images.TestBlendImages)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/opt/conda/lib/python3.6/site-packages/parameterized/parameterized.py", line 533, in standalone_func

    return func(*(a + p.args), **p.kwargs)

  File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/test_blend_images.py", line 39, in test_blend

    blended = blend_images(image[None], label[None])

  File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/visualize/utils.py", line 144, in blend_images

    label = rescale_array(label)

  File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/transforms/utils.py", line 167, in rescale_array

    norm = (arr - mina) / (maxa - mina)  # normalize the array first

RuntimeError: Integer division of tensors using div or / is no longer supported, and in a future release div will perform true division as in Python 3. Use true_divide or floor_divide (// in Python) instead.



======================================================================

ERROR: test_blend_5 (tests.test_blend_images.TestBlendImages)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/opt/conda/lib/python3.6/site-packages/parameterized/parameterized.py", line 533, in standalone_func

    return func(*(a + p.args), **p.kwargs)

  File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/test_blend_images.py", line 39, in test_blend

    blended = blend_images(image[None], label[None])

  File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/visualize/utils.py", line 144, in blend_images

    label = rescale_array(label)

  File "/home/jenkins/agent/workspace/MONAI-premerge/monai/monai/transforms/utils.py", line 167, in rescale_array

    norm = (arr - mina) / (maxa - mina)  # normalize the array first

RuntimeError: Integer division of tensors using div or / is no longer supported, and in a future release div will perform true division as in Python 3. Use true_divide or floor_divide (// in Python) instead.



----------------------------------------------------------------------

```

**Pull request merged**